docs/proposals/14-adminkubeconfig-subresource.md

```diff
@@ -0,0 +1,103 @@
+gep-number: 14
+creation-date: 2021-04-02
+status: implementable
+authors:
+- "@mvladev"
+reviewers:
+- "@rfranzke"
+---
+
+# GEP-14: Dynamic kubeconfig generation for Shoot clusters
+
+## Table of Contents
+
+- [GEP-14: Dynamic kubeconfig generation for Shoot clusters](#gep-14-dynamic-kubeconfig-generation-for-shoot-clusters)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+  - [Alternatives](#alternatives)
+
+## Summary
+
+This `GEP` introduces new  `Shoot` subresource called `AdminKubeConfigRequest` allowing for users to dynamically generate a short-lived `kubeconfigs` that can be used to access the `Shoot` cluster as `cluster-admin`.
+
+## Motivation
+
+Today, when access to the created `Shoot` clusters is needed, a `kubeconfig` with static token credentials is used. This static token is in the `system:masters` group, granting it `cluster-admin` privileges. The `kubeconfig` is generated when the cluster is reconciled, stored in `ShootState` and replicated in the `Project`'s namespace in a `Secret`. End-users can fetch the secret and use the `kubeconfig` inside it.
+
+There are several problems with this approach:
+
+- The token in the `kubeconfig` does not have any expiration, so end-users have to issue request a `kubeconfig` credential rotation call manually.
+- There is no user identity in the token. e.g. if user `Joe` gets the `kubeconfig` from the `Secret`, user in that token would be `system:cluster-admin` and not `Joe` when accessing the `Shoot` cluster with it. This makes auditing events in the cluster almost impossible.
+
+### Goals
+
+- Add a `Shoot` subresource called `adminkubeconfig` that would produce a `kubeconfig` used to access that `Shoot` cluster.
+- The `kubeconfig` is not stored in the API Server, but generated for each request.
+- In the `AdminKubeConfigRequest` send to that subresource, end-users can specify the expiration time of the credential.
+- The identity (user) in the Gardener cluster would be part of the identity (x509 client certificate). E.g if `Joe` authenticates against the Gardener API server, the generated certificate for `Shoot` authentication would have the following subject:
+
+  - Common Name: `Joe`
+  - Organisation: `system:masters`
+
+- The maximum validity of the certificate can be enforced by setting a flag on the `gardener-apiserver`.
+- Deprecate and remove the old `{shoot-name}.kubeconfig` secrets in each `Project` namespace.
+
+### Non-Goals
+
+- Generate `OpenID Connect` kubeconfigs
+
+## Proposal
+
+The `gardener-apiserver` would serve a new `shoots/adminkubeconfig` resource. It can only accept `CREATE` calls and accept `AdminKubeConfigRequest`. A `AdminKubeConfigRequest` would have the following structure:
+
+```yaml
+apiVersion: authentication.gardener.cloud/v1alpha1
+kind: AdminKubeConfigRequest
+spec:
+  expirationSeconds: 3600
```

Comment:
I am fine that OpenID is not a goal for this GEP, but it would be nice to implement the API in a way that it can easily be extended to support this case in the future.

```suggestion
spec:
  type: client-ca # In the future, a new type `oidc` can be supported
  expirationSeconds: 3600
```

Reply:
I would have a different subresource for OIDC later, because the semantics there is completely different.